clients/lsp-netbeans.el

```diff
@@ -0,0 +1,129 @@
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+;;; Commentary:
+
+;; LSP and DAP client for the Apache Netbeans LSP server.
+
+;; Currently missing many features that are in the VSCode extension, including, but not limited to:
+;; - All custom commands, such as
+;;     java.build.workspace
+;;     java.generate.getters
+;;     java.generate.setters
+;;     java.generate.getters.setters
+;;     java.generate.implementMethod
+;;     java.generate.overrideMethod
+;;     java.generate.equals.hashCode
+;;     java.generate.delegateMethod
+;;     java.generate.constructor
+;;     java.generate.logger
+;;     java.generate.toString
+;;     graalvm.pause.script
+;;     java.super.implementation
+;; - Groovy support
+;; - Managing/running/debugging unittest directly
+;; - GraalVM Native Image launch/debug configuration
+;; - Attach to JVM process debug configs
+;; - Launch JVM debug config
+
+;;; Code:
+
+(require 'lsp-mode)
+(require 'dap-mode)
+
+(defcustom lsp-netbeans-download-url "https://marketplace.visualstudio.com/_apis/public/gallery/publishers/ASF/vsextensions/apache-netbeans-java/12.4.301/vspackage"
+  "URL to download the Apache Netbeans LSP server from"
+  :group 'lsp-netbeans
+  :type 'string)
+
+(defcustom lsp-netbeans-install-dir (f-join lsp-server-install-dir "asf.apache-netbeans-java")
+  "Apache Netbeans language server installation dir"
+  :group 'lsp-netbeans
+  :type 'string)
+
+(defun lsp-netbeans-server-command (main-port)
+  `(,(f-join lsp-netbeans-install-dir "run.sh")
+    "--start-java-debug-adapter-server=listen:0"
+    ,(format "--start-java-language-server=listen:%d" main-port)))
+
+(defun lsp-netbeans--install-server (client callback error-callback update?)
+  (unless (and (not update?)
+               (f-exists? lsp-netbeans-install-dir))
+    (make-directory lsp-netbeans-install-dir t)
+    (let ((download-path (f-join lsp-netbeans-install-dir "vspackage")))
+      (lsp-download-install
+       (lambda ()
+         (let ((run-script-path (f-join lsp-netbeans-install-dir "run.sh")))
+           (with-temp-file run-script-path
+             (insert "#!/bin/bash
+                      cd `dirname $0`
+                      cd extension
+                      node out/nbcode.js $@ >&2")
+             (executable-make-buffer-file-executable-if-script-p)))
+         (message "Done downloading Netbeans LSP server"))
+       (lambda (err) (error err))
+       :url lsp-netbeans-download-url
+       :store-path download-path
+       :decompress :zip))))
+
+(lsp-register-client
+ (make-lsp-client
+  :new-connection (lsp-tcp-connection 'lsp-netbeans-server-command)
+  :activation-fn (lsp-activate-on "java")
+  :server-id 'netbeans
+  :initialization-options (list :nbcodeCapabilities
+                                (list
+                                 :testResultsSupport (lsp-json-bool nil)
```

Comment:
These three should probably be settings, as in the VSCode extension, but I'm not sure what they even imply.